services/oauth/v2/controlplane/cp_connector.go

```diff
@@ -0,0 +1,183 @@
+		c.client = client
+	}
+}
+
+// WithLogger is a functional option to set the parent logger for the Connector
+func WithLogger(parentLogger logger.Logger) func(*connector) {
+	return func(c *connector) {
+		c.logger = parentLogger
+	}
+}
+
+// WithCpClientTimeout is a functional option to set the timeout for the Connector
+func WithCpClientTimeout(timeout time.Duration) func(*connector) {
+	return func(c *connector) {
+		c.timeout = timeout
+	}
+}
+
+// processResponse is a helper function to process the response from the control plane
+func processResponse(resp *http.Response) (statusCode int, respBody string) {
+	var respData []byte
+	var ioUtilReadErr error
+	if resp != nil && resp.Body != nil {
+		respData, ioUtilReadErr = io.ReadAll(resp.Body)
+		if ioUtilReadErr != nil {
+			return http.StatusInternalServerError, ioUtilReadErr.Error()
+		}
+	}
+	// Detecting content type of the respData
+	contentTypeHeader := strings.ToLower(http.DetectContentType(respData))
+	// If content type is not of type "*text*", overriding it with empty string
+	if !contentTypePattern.MatchString(contentTypeHeader) {
+		respData = []byte("")
+	}
+
+	return resp.StatusCode, string(respData)
+}
+
+// CpApiCall is a function to make a call to the control plane, handle the response and return the status code and response body
+func (c *connector) CpApiCall(cpReq *Request) (int, string) {
+	cpStatTags := stats.Tags{
+		"url":          cpReq.URL,
+		"requestType":  cpReq.RequestType,
+		"destType":     cpReq.DestName,
+		"method":       cpReq.Method,
+		"flowType":     string(cpReq.rudderFlowType),
+		"oauthVersion": "v2",
+	}
+
+	var reqBody *bytes.Buffer
+	var req *http.Request
+	var err error
+	if cpReq.Body != "" {
+		reqBody = bytes.NewBufferString(cpReq.Body)
+		req, err = http.NewRequest(cpReq.Method, cpReq.URL, reqBody)
+	} else {
+		req, err = http.NewRequest(cpReq.Method, cpReq.URL, http.NoBody)
+	}
+	if err != nil {
+		c.logger.Errorn("[request] :: destination request failed",
+			logger.NewErrorField(err))
+		// Abort on receiving an error in request formation
+		return http.StatusBadRequest, err.Error()
+	}
+	// Authorisation setting
+	req.SetBasicAuth(cpReq.BasicAuthUser.BasicAuth())
+
+	// Set content-type in order to send the body in request correctly
+	if cpReq.ContentType != "" {
+		req.Header.Set("Content-Type", cpReq.ContentType)
+	}
+
+	cpApiDoTimeStart := time.Now()
+	res, doErr := c.client.Do(req)
+	defer func() { httputil.CloseResponse(res) }()
+	c.stats.NewTaggedStat("cp_request_latency", stats.TimerType, cpStatTags).SendTiming(time.Since(cpApiDoTimeStart))
+	c.logger.Debugn("[request] :: destination request sent")
+	if doErr != nil {
+		// Abort on receiving an error
+		c.logger.Errorn("[request] :: destination request failed",
+			logger.NewErrorField(doErr))
+		errorType := GetErrorType(doErr)
+		cpStatTags["errorType"] = errorType
+		c.stats.NewTaggedStat("oauth_v2_cp_request_error", stats.CountType, cpStatTags).Count(1)
```

Comment:
I am personally in favour of capturing both errors and non-errors on the same metric, using diffrent tags. This makes it very easy and accurate to compute things like failure rate.

I would suggest a metric like `oauth_v2_cp_requests` renaming `errorType` to `error` and capture `error`=none incase there is no error.

Again this is just a preference, I don't have a strong argument against this approach, you can use the latency metric above for the total count of requests. It just might take some time for someone exploring the metrics to figure it out.

Reply:
Addressed for some of the stats, will take this as an improvement in follow-up PRs